src/kernel/vmm.zig

```diff
@@ -0,0 +1,543 @@
+const build_options = @import("build_options");
+const mock_path = build_options.mock_path;
+const builtin = @import("builtin");
+const is_test = builtin.is_test;
+const std = @import("std");
+const bitmap = @import("bitmap.zig");
+const pmm = @import("pmm.zig");
+const mem = if (is_test) @import(mock_path ++ "mem_mock.zig") else @import("mem.zig");
+const tty = @import("tty.zig");
+const multiboot = @import("multiboot.zig");
+const log = @import("log.zig");
+const panic = @import("panic.zig").panic;
+
+/// Attributes for a virtual memory allocation
+pub const Attributes = struct {
+    /// Whether this memory belongs to the kernel and can therefore not be accessed in user mode
+    kernel: bool,
+
+    /// If this memory can be written to
+    writable: bool,
+
+    /// If this memory can be cached. Memory mapped to a device shouldn't, for example
+    cachable: bool,
+};
+
+/// All data that must be remembered for a virtual memory allocation
+const Allocation = struct {
+    /// The physical blocks of memory associated with this allocation
+    physical: std.ArrayList(usize),
+};
+
+///
+/// Returns a container that can map and unmap virtual memory to physical memory.
+/// The mapper can pass some payload data when mapping an unmapping, which is of type `Payload`. This can be anything that the underlying mapper needs to carry out the mapping process.
+///
+/// Arguments:
+///     IN comptime Payload: type - The type of the VMM-specific payload to pass when mapping and unmapping
```

Comment:
Could explain a bit more what a payload is as I had to look for it in the arch file.

Reply:
OK I've added some more clarification. Let me know if it needs improving.